networkx/generators/directed.py

```diff
@@ -238,15 +238,13 @@ def scale_free_graph(
            Discrete Algorithms, 132--139, 2003.
     """
 
-    def _choose_node(G, distribution, delta, psum):
-        cumsum = 0.0
-        # normalization
-        r = seed.random()
-        for n, d in distribution:
-            cumsum += (d + delta) / psum
-            if r < cumsum:
-                break
-        return n
+    def _choose_node(candidates, n_nodes, delta):
+        if delta > 0:
+            bias_sum = n_nodes * delta
+            p_delta = bias_sum / (bias_sum + len(candidates))
+            if seed.random() < p_delta:
+                return seed.randint(0, n_nodes)
```

Comment:
Implicit here is an assumption that the nodes are essentially drawn from `range(0, n_nodes)`. For cases where this isn't true, the returned nodes might seem strange to users:

```python
>>> G = nx.scale_free_graph(10)
>>> H = nx.relabel_nodes(G, {i: 2**i for i in range(10)}
>>> H.nodes()
NodeView((1, 2, 4, 8, 16, 32, 64, 128, 256, 512))
>>> K = nx.scale_free_graph(20, create_using=H, delta_in=1e6, delta_out=1e6)
>>> K.nodes()
NodeView((1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 7, 11, 12, 5, 14, 0, 9, 13, 18, 17))
```

AFAICT this doesn't affect the algorithm behavior so it isn't wrong. However it might be worth just adding a note to the docstring about how new nodes are selected so that users starting with an initial graph that has non-integer nodes or unevenly spaced integer nodes know what to expect.

Reply:
Perhaps we can use that random number as an index to the candidates list so it returns a node instead of an integer?  But I might be misunderstanding the intent....
```return candidate[seed.randint(0, n_nodes)]```

Reply:
Great catch! I've added a commit that properly supports working with existing graphs by introducing a node state (simple list of nodes) and a cursor for new nodes that starts at the highest already occupied `int` + 1.

The code would be simpler if we would sample from `list(G.nodes())` but it would make the code orders of magnitude slower once again for large `n`. The current approach does not sacrifice any of the performance gains, timing results are still similar to posted above.

@rossbar your snippet now produces
```python
NodeView((1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 513, 514, 515, 516, 517, 518, 519, 520, 521, 522))
```

But it now also supports e.g. strings:
```python
>>> G = nx.scale_free_graph(10)
>>> H = nx.relabel_nodes(G, {i: str(2**i) for i in range(10)})
>>> H.nodes()
NodeView(('1', '2', '4', '8', '16', '32', '64', '128', '256', '512'))
>>> K = nx.scale_free_graph(20, create_using=H, delta_in=1e6, delta_out=1e6)
>>> K.nodes()
NodeView(('1', '2', '4', '8', '16', '32', '64', '128', '256', '512', 0, 1, 2, 3, 4, 5, 6, 7, 8, 9))
```

@dschult your approach unfortunately wouldn't work because `len(candidates)` > `len(nodes)`, as `candidates` contains the nodes repeated `d` times where `d` are their degrees. Also, sampling from `candidates` brings us back to sampling with weights determined by the degrees, and in that piece of the code we want to sample uniformly.